common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java

```diff
@@ -218,6 +242,45 @@ public static void load(String originalName, ClassLoader loader) {
         }
     }
 
+    /**
+     * Try to patch shaded library to ensure it uses an unique id.
+     */
+    private static void patchShadedLibraryId(byte[] bytes, String originalName, String name) {
+        // Our native libs always have the name as part of their id so we can search for it and replace it
+        // to make the id unique if shading is used.
+        byte[] nameBytes = originalName.getBytes(CharsetUtil.UTF_8);
+        int idIdx = -1;
+
+        // Be aware this is a really raw way of patching a dylib but it does all we need without implementing
+        // a full mach-o parser and writer. Basically we just replace the the original bytes with some
+        // random bytes as part of the id regeneration. The important thing here is that we need to use the same
+        // length to not corrupt the mach-o header.
+        outerLoop: for (int i = 0; i < bytes.length && bytes.length - i >= nameBytes.length; i++) {
+            int idx = i;
+            for (int j = 0; j < nameBytes.length;) {
+                if (bytes[idx++] != nameBytes[j++]) {
+                    // Did not match the name, increase the index and try again.
+                    break;
+                } else if (j == nameBytes.length) {
+                    // We found the index within the id.
+                    idIdx = i;
+                    break outerLoop;
+                }
+            }
+        }
+
+        if (idIdx == -1) {
+            logger.debug("Was not able to find the id of the shaded native library {}, can't adjust it.", name);
+        } else {
+            // we found our id... now monkey-patch it!
+            logger.debug(
+                    "Found the id of the shaded native library {}, adjusting it to something unique.", name);
+
+            PlatformDependent.threadLocalRandom().nextBytes(nameBytes);
```

Comment:
Perhaps we should limit the characters to alphabets and numbers and log the new name just in case it can be used for debugging? We might also want to preserve the first few characters?

Reply:
@trustin this is actually only part of the id as the id usually contains the full path of the native lib when it was compiled. So it does not really make sense to log. That said I think limit to numbers + alphabets is a good idea.